Second iteration of the proposal

aspiwack opened the pull request:
Trying to address all the points so far.

aspiwack committed:
Added section to be filled, in order not to forget them

aspiwack committed:
A section on exceptions and catch

aspiwack committed:
Linear exceptions

aspiwack committed:
Placeholder for discussion on affine types

aspiwack committed:
WIP affine types

aspiwack committed:
Still WIPing

aspiwack committed:
WIP

aspiwack committed:
Add some internal links

aspiwack committed:
Paragraph on additive conjunction

aspiwack committed:
Typing of records and projections

aspiwack committed:
Unresolved pattern-related things

aspiwack committed:
Remove 0 from the main body of the proposal

And move to alternatives

aspiwack committed:
Remove misleading aside

aspiwack committed:
(->.) vs (-o)

aspiwack committed:
A todo note

aspiwack committed:
Unboxed data type

aspiwack committed:
Revamped introduction

- Emphasis on the fact that all these examples are applications of the
  proposal, and not part of the proposal
- Numerous references to the paper

aspiwack committed:
Insist on abstraction vs changes to RTS, etc…

aspiwack committed:
Replace IOL with RIO

aspiwack committed:
Be more explicit on the guarantees in presence of exception

aspiwack committed:
Clarification on catch and throw

aspiwack committed:
Links in affine vs linear

aspiwack committed:
Propose to keep affine types for later

aspiwack committed:
Section on more multiplicities

aspiwack committed:
Misc clarification

aspiwack committed:
Link to linear-base

aspiwack committed:
Split the former formalism section into smaller sections

aspiwack committed:
Fix heading levels

aspiwack committed:
More link to the paper

aspiwack committed:
Deep patterns & multiple equations

aspiwack committed:
A few sentences on inference

aspiwack committed:
Effects on Core

aspiwack committed:
Additional cross-reference

aspiwack committed:
Pointer to Tov's thesis

aspiwack committed:
Discussion of encodings

mboes committed:
Edits to motivation section.

Removed all examples. Because now that we can lean on the paper, let's
just point to the examples in the paper as written there.

aspiwack commented on proposals/0000-linear-types.rst:
This isn't in the paper either (though I believe it is in a blog post).

aspiwack commented on proposals/0000-linear-types.rst:
I don't think we have this in the paper (I made it part of the Haskell Exchange talk, though).

aspiwack commented on proposals/0000-linear-types.rst:
Missing is typestate.

aspiwack reviewed (commented):
Just a note to better source the listed examples.

mboes committed:
misc edits.

aspiwack commented on proposals/0000-linear-types.rst:
Note for later: « hard to impossible to ».

(I don't think that they are impossible though. With a good, unpalatable, application of indexed monad, you can probably enforce such invariants)

mboes committed:
Rewrite affine types comparison section.

aspiwack commented on proposals/0000-linear-types.rst:
```diff
@@ -903,49 +903,86 @@ cope with affine functions. Therefore the choice between these three
+roots do get deleted eventually (at scope exit), but in practice, in
+complex dual-language projects, introducing neither too fine-grained
+or too coarse-grained scopes has proven very difficult. Furthermore,
+``bracket``-like constructs break tail-recursion. Linear types enable
```
tail-recursion: good point.

aspiwack commented on proposals/0000-linear-types.rst:
```diff
@@ -903,49 +903,86 @@ cope with affine functions. Therefore the choice between these three
+``bracket``-like constructs break tail-recursion. Linear types enable
+working with a single global resource scope, while still guaranteeing
+eventual deletion of roots, in any order. Affine types do not. At any
+rate, not in direct-style.
```
This paragraph is a tad long. Maybe we ought to par down the most technical details, and defer them to the blog post (following the mantra that the blog post has more space and spends quite a bit of space and effort into making the explanation clear).

mboes replied:
We could, but my concern is that the blog post actually does not explain in a single pithy paragraph why multiple scopes are so bad. If the blog post did have that, then I'd be happy to link to that instead. That said, I found a couple of sentences to cut out, so I'll do that.

aspiwack commented on proposals/0000-linear-types.rst:
```diff
@@ -903,49 +903,86 @@ cope with affine functions. Therefore the choice between these three
+roots are always freed exactly once, so we must register each GC root
+to free them if an exception is thrown. A free-at-most-once guarantee
+wouldn't require this. The design cost is that ``catch`` requires
+a weaker type than desirable, as discussed above, limiting its power.
```
I'm not sure I follow the point here.

First, on a purely technical stand-point: if I understand correctly we don't do any registering, Java does (with `pushLocalFrame`, `popLocalFrame` module the actual names).

Second, even if we had to do the registration thing, affine types would no alleviate that cost (if anything, it would make it more acute). It's just that it would apply both in case of normal exit and of exception. The point about `catch` still stands though.

mboes replied:
In our particular case it's the JNI that bears this implementation cost. So I could say there's actually no cost here, but in general that's not true, so I let it slide. Good point about affine types needing same in practice, even if it's not needed to comply with the protocol the types demand. I'll have to reword.

aspiwack commented on proposals/0000-linear-types.rst:
```diff
@@ -962,7 +999,7 @@ example:
   data Closed (n :: *)
 
-  socket :: RIO (forall n. Socket n 'Unbound :'A-> RIO (Unrestricted a, Closed s)) :'A -> RIO (Unrestricted a)
+  newSocket :: RIO (forall n. Socket n 'Unbound :'A-> RIO (Unrestricted a, Closed s)) :'A -> RIO (Unrestricted a)
```
Note that I called it `socket` because it's [the name](https://www.stackage.org/haddock/lts-9.18/network-2.6.3.2/Network-Socket.html#v:socket) of this primitive in the socket library.

aspiwack reviewed (commented):
I like these changes, though I think we should work and compress the JNI story a bit.

mboes committed:
Shorten JNI text somewhat.

mboes committed:
Note that need registration too even with affine types.

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -469,56 +501,225 @@ understood as the latter, well-typed, one. It means that is not a
-  -- Not linear! But well-typed if the above is accepted
-  f :: [a] ->. (Int, [a])
-  f l = (badLength l, l)
+are linear constructors: ``R :: A1 ->. … ->. An ->. R``. Projections
```
missing subject?

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -469,56 +501,225 @@ understood as the latter, well-typed, one. It means that is not a
+Because of backwards compatibility, we initially chose the following
+strategy: when the type of a function is not constrained by given
+constraints, we infer it to have multiplicity ω.
+
```
infer -> conservatively assume

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -469,56 +501,225 @@ understood as the latter, well-typed, one. It means that is not a
+<https://github.com/tweag/linear-base/blob/master/src/System/IO/Resource.hs>`_
+for a detailed implementation).
+
+First, note that since Haskell program are of type ``IO ()``, we need a
```
programs

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -469,56 +501,225 @@ understood as the latter, well-typed, one. It means that is not a
+
+If no exception occurs, then all resources have been released by the
+program. In case an exception occurs, the program jumps to ``runRIO``,
+which releases the leftover resources.
```
jumps to a handler in runRIO ? Unclear.

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -469,56 +501,225 @@ understood as the latter, well-typed, one. It means that is not a
+
+  throwRIO :: Exception e => e -> RIO a
+  trow :: Exception e => e -> a
+
```
typo: throw

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -642,6 +846,189 @@ except to the left of an arrow. And ``WithMult a p -> b`` means
+Binders with multiplicity
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the paper, we wrote ``λ x :₁ A, u`` for (unannotated) linear
```
. for ,

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -642,6 +846,189 @@ except to the left of an arrow. And ``WithMult a p -> b`` means
+not necessarily consume their arguments. For instance, an ``RIO a``
+may abort before closing its file handles. And because of ``catch`` we
+are able to be observe this effect. Could affine types jibe better
+with this reality?
```
cope for jibe

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -469,56 +501,225 @@ understood as the latter, well-typed, one. It means that is not a
+	 -> (e -> RIO (Unrestricted a))
+	 -> RIO (Unrestricted a)
+
+That is: no linear resource previously allocated can be referenced in
```
As far as I can see, you can make the body linear. And so, the linear variables can occur in it. If a variable turns out not to be used before an exception occurs it can be freed with the mechanism described above. Am I missing something?

aspiwack replied:
The problem here is that you have to know how to free the linear thing. Which you usually do not. The mechanism described above works because there is a cooperation between `RIO` and the resource types it manages. It can't, in general, be assumed of linear variable that they cooperate this way.

Having a linear catch this way would make it possible to break abstractions enforced by linear types.

jyp replied:
Alright, I was assuming that all linear resources were allocated using RIO. 

Perhaps you could add a line saying "because there could be linear resources coming from outside RIO".

aspiwack replied:
The truth is, though, that even if the linear resources are allocated by RIO, we would still need to figure out, in implementation of `catch`, which resources were captured by the body and must be deallocated by `catch` in case of exception. I don't know how to do this either. Otherwise, the resource would be held dangling until the entire `runRIO` scope is finished. Which may be a long time.

But this is less of a logic problem and more of an implementation detail. I suppose.

jyp replied:
I thought that this is what RIO does (closing resources only when the entire RIO finishes).

aspiwack replied:
Well, not quite: they are closed when a deallocating primitive is called. But if there is an exception, the entire `RIO` aborts, and frees remaining resources.

So the point is that you tend to have one big `runRIO` scope, and you do your resource management in it using linear types.

But if we start catching exceptions carelessly in the middle of an `RIO`, we may lose the ability to manage some resources, effectively leaking resources.

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -642,6 +846,189 @@ except to the left of an arrow. And ``WithMult a p -> b`` means
+
+Finally, while it is easy to implement system (3), we have not
+included it in the proposal. We propose to shelve it for a later
+proposal (see also `More multiplicities`_ below), while thriving in
```
'shelve' is often used colloquially for ignore indefinitely. Prefer 'reserve'

jyp commented on proposals/0000-linear-types.rst:
```diff
@@ -448,6 +394,92 @@ Then ``xi`` has multiplicity annotation ``p*qi``. For instance
+
+Unresolved questions
+++++++++++++++++++++
+
```
I am not sure at all what the point is here. If there is an issue with the translation to Core, then the translated code need to be shown. Furthermore, I am not convinced that there is any significant issue. (I wrote an email on Nov. 21 2017 about a simple solution).

aspiwack committed:
Fix heading levels and add section numbering

aspiwack committed:
@jyp's remarks

aspiwack committed:
Add link to manually-managed data

aspiwack committed:
Add an anchor

will use later

aspiwack committed:
Remove '1 and 'U abbreviations

aspiwack committed:
Position anchors below the headings

Otherwise, Github gets confused and the anchor is not created. Causing
internal links to be dead.

aspiwack committed:
Formatting

aspiwack committed:
Subheadings for readability

aspiwack committed:
More fixing links

aspiwack committed:
Formatting

aspiwack committed:
More link fixing

aspiwack committed:
Minor improvement, proposed by @simonpj

aspiwack committed:
Alternative notations for the mulitplicity-parametric arrow

As suggested by @simonpj

aspiwack committed:
Better syntax specification

- What is really syntax vs what is just recognised specially
- More explicit description of the grammare for `:p->`
    - With an alternative for the latter

aspiwack committed:
Data type declarations without -XLinearTypes (part 1)

aspiwack committed:
More link fixes

aspiwack committed:
Data type declarations without -XLinearTypes (part 2)

aspiwack committed:
Illustration: compiling pattern matching may yield ill-typed terms

aspiwack committed:
Precision on RIO's catch function
